### PR TITLE
feat: transition to binary serialization

### PR DESCRIPTION
**Pull request by delehef**
Drop JSON to use a binary serialization format to (i) reduce memory usage and (ii) improve serialization & deserialization performances.

### CODE REVIEW

File: arithmetization/src/main/java/net/consensys/linea/tracegeneration/rpc/RollupGenerateConflatedTracesToFileV0.java [hunk 1]
```diff
@@ -121,52 +118,26 @@ private TraceService initTraceService() {
...
-        .toFile();
+    return Paths.get(
+        String.format(
+            "%.10s-%s.traces.%s",
+            System.currentTimeMillis(), tracesEngineVersion, getFileFormat()));
   }
 
   private String getFileFormat() {
-    return isGzipEnabled ? "json.gz" : "json";
+    return "lt";
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/tracegeneration/rpc/RollupGenerateConflatedTracesToFileV0.java:141]: Is `lt` [this](https://www.analog.com/en/technical-articles/lt-programming-hex-file-format-documentation-in-circuit-programming.html) ?

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java:1]: Not really following what's going on here. Is this information for the binary format to know how long a column will be in storage ?

Comment (OlivierBBB) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java:1]: I guess it's something to that effect, considering [this](arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java)

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+package net.consensys.linea.zktracer;
+
+public record ColumnHeader(String name, int bytesPerElement, int length) {
+  public int dataSize() {
+    return this.length() * this.bytesPerElement();
+  }
+
+  public int headerSize() {
+    return 2
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java:24]: What is the `2` in the formula ?

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java [hunk 3]
```diff
@@ -54,11 +63,53 @@ public ZkTracer() {
...
+
+  @Override
+  public void writeToFile(final Path filename) {
+    log.warn("[TRACING] Starting serialization to " + filename.toAbsolutePath());
+    Stopwatch sw = Stopwatch.createStarted();
+
+    final List<Module> modules = this.hub.getModulesToTrace();
+    final List<ColumnHeader> traceMap =
+        modules.stream().flatMap(m -> m.columnsHeaders().stream()).toList();
+    final int headerSize = traceMap.stream().mapToInt(ColumnHeader::headerSize).sum() + 4;
```
Comment (OlivierBBB) [inline arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java:84]: What's that `+4` ?

**Review by OlivierBBB — COMMENTED** [review]
I don't understand a lot of what's going on but AFAICT there should be no substantial changes for module implementers. LGTM

**Review by letypequividelespoubelles — APPROVED** [review]

File: arithmetization/src/main/java/net/consensys/linea/tracegeneration/rpc/RollupGenerateConflatedTracesToFileV0.java (hunk 1 above)
Comment (delehef) [reply arithmetization/src/main/java/net/consensys/linea/tracegeneration/rpc/RollupGenerateConflatedTracesToFileV0.java:141]: Nothing so deliberate, `lt` is **L**inea **T**trace :)

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java
Comment (delehef) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java:1]: Exactly

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java (hunk 2 above)
Comment (delehef) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java:24]: 2 bytes for the length of the string (`// i16: name size`) but spotless puts the comment on another line.

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java (hunk 3 above)
Comment (delehef) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java:84]: An i32 for the number of columns to read from the trace

```
header.putInt(traceMap.size());
```

File: arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java
Comment (OlivierBBB) [reply arithmetization/src/main/java/net/consensys/linea/zktracer/ColumnHeader.java:1]: Question: how does this prediction gel with perspectives ? Where a single "physical column" is the merger of several "virtual columns" ? Do these columns exist in some way in the trace ?

**Pull request merged**